include/drivers/gpio.h

```diff
@@ -304,6 +304,57 @@ typedef uint8_t gpio_dt_flags_t;
  */
 typedef uint32_t gpio_flags_t;
 
+/**
+ * @brief Provides a type to hold GPIO information specified in devicetree
+ *
+ * This type is sufficient to hold a GPIO device pointer, pin number,
+ * and the subset of the flags used to control GPIO configuration
+ * which may be given in devicetree.
+ */
+struct gpio_dt_spec {
+	const struct device *gpio;
+	gpio_pin_t pin;
+	gpio_dt_flags_t dt_flags;
+};
+
+/**
+ * @brief Static initializer for a @p gpio_dt_spec
+ *
+ * This returns a static initializer for a @p gpio_dt_spec structure
+ * given a devicetree node identifier and a property specifying a
+ * GPIO.
+ *
+ * Example devicetree fragment:
+ *
+ *     n: node {
+ *             foo-gpios = <&gpio1 2 GPIO_ACTIVE_LOW>;
+ *     }
+ *
+ * Example usage:
+ *
+ *     GPIO_DT_SPEC_GET(DT_NODELABEL(n), foo_gpios)
+ *     // {
+ *     //         .gpio = DEVICE_DT_GET(DT_NODELABEL(n)),
+ *     //         .pin = 2,
+ *     //         .dt_flags = GPIO_ACTIVE_LOW
+ *     // }
+ *
+ * The 'gpio' field must still be checked for readiness, e.g. using
+ * device_is_ready(). It is an error to use this macro unless the node
+ * exists, has the given property, and that property specifies a GPIO
+ * controller, pin number, and flags as shown above.
+ *
+ * @param node_id devicetree node identifier
+ * @param prop lowercase-and-underscores property name
+ * @return static initializer for a struct gpio_dt_spec for the property
+ */
+#define GPIO_DT_SPEC_GET(node_id, prop)					\
+	{								\
+		.gpio = DEVICE_DT_GET(DT_GPIO_CTLR(node_id, prop)),	\
+		.pin = DT_GPIO_PIN(node_id, prop),			\
+		.dt_flags = DT_GPIO_FLAGS(node_id, prop),		\
+	}
+
```

Comment:
Seems like helpers for `gpio_dt_pin_configure(gpio_dt_spec, flags)` and `gpio_dt_pin_interrupt_configure(gpio_dt_spec, flags)` would be useful.

Reply:
> Seems like helpers for `gpio_dt_pin_configure(gpio_dt_spec, flags)` and `gpio_dt_pin_interrupt_configure(gpio_dt_spec, flags)` would be useful.

Good idea. I named these gpio_pin_configure_dt and gpio_pin_interrupt_configure_dt instead because I think @pabigot will ask for that to make it easy to grep for calls to either the DT or non-DT versions.

Reply:
I find the functions in @galak's proposal easier to read. Also we have already `gpio_dt_flags_t`, `gpio_dt_spec`, `GPIO_DT_SPEC_GET` so the `gpio_dt_` prefix seems to be well established. That said, I don't have strong feelings about that.